File: backend/app/api/v1/patients.py
from typing import Annotated, Optional
import uuid

from fastapi import APIRouter, Depends, Query, UploadFile, File
from sqlalchemy.ext.asyncio import AsyncSession
import redis.asyncio as aioredis

from app.core.database import get_db
from app.core.dependencies import get_redis, require_roles, get_current_user, require_doctor, require_admin
from app.models.user import User, UserRole
from app.schemas.patients import (
    PatientCreate, PatientUpdate, PatientResponse,
    AllergyCreate, AllergyUpdate, AllergyResponse,
    ChronicDiseaseCreate, ChronicDiseaseUpdate, ChronicDiseaseResponse,
    MedicalCardUpdate, MedicalCardResponse,
    DocumentResponse, EncounterSummary,
    AllergenResponse,
)
from app.services.patient_service import PatientService

router = APIRouter(prefix="/patients", tags=["Patients"])

# ─── Role dependencies ────────────────────────────────────────────────────────

_doctor_admin_patient = require_roles(
    UserRole.DOCTOR, UserRole.ADMIN, UserRole.SUPER_ADMIN, UserRole.PATIENT
)
_admin_only = require_roles(UserRole.ADMIN, UserRole.SUPER_ADMIN)


# ─── Patients ─────────────────────────────────────────────────────────────────

@router.post("", response_model=PatientResponse, status_code=201)
async def create_patient(
    data: PatientCreate,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    return await PatientService(db, redis).create_patient(data, created_by_user_id=current_user.id)


@router.get("", response_model=list[PatientResponse])
async def list_patients(
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
    search: Optional[str] = Query(None, description="Search by name, tax_id, or phone"),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=200),
):
    return await PatientService(db, redis).get_patients(search=search, skip=skip, limit=limit)


@router.delete("/{patient_id}", status_code=204)
async def deactivate_patient(
    patient_id: uuid.UUID,
    current_user: Annotated[User, Depends(_admin_only)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    """Деактивувати пацієнта без облікового запису (soft delete). Тільки ADMIN."""
    await PatientService(db, redis).deactivate_patient(patient_id, current_user)


@router.get("/{patient_id}", response_model=PatientResponse)
async def get_patient(
    patient_id: uuid.UUID,
    current_user: Annotated[User, Depends(_doctor_admin_patient)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    return await PatientService(db, redis).get_patient(patient_id, requesting_user=current_user)


@router.patch("/{patient_id}", response_model=PatientResponse)
async def update_patient(
    patient_id: uuid.UUID,
    data: PatientUpdate,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    return await PatientService(db, redis).update_patient(patient_id, data, updated_by=current_user)


# ─── Medical card ─────────────────────────────────────────────────────────────

@router.get("/{patient_id}/medical-card", response_model=MedicalCardResponse)
async def get_medical_card(
    patient_id: uuid.UUID,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    return await PatientService(db, redis).get_medical_card(patient_id, requesting_user=current_user)


@router.put("/{patient_id}/medical-card", response_model=MedicalCardResponse)
async def update_medical_card(
    patient_id: uuid.UUID,
    data: MedicalCardUpdate,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    return await PatientService(db, redis).update_medical_card(patient_id, data, updated_by=current_user)


# ─── Allergies ────────────────────────────────────────────────────────────────

@router.post("/{patient_id}/allergies", response_model=AllergyResponse, status_code=201)
async def add_allergy(
    patient_id: uuid.UUID,
    data: AllergyCreate,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    return await PatientService(db, redis).add_allergy(patient_id, data)


@router.patch("/{patient_id}/allergies/{allergy_id}", response_model=AllergyResponse)
async def update_allergy(
    patient_id: uuid.UUID,
    allergy_id: uuid.UUID,
    data: AllergyUpdate,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    return await PatientService(db, redis).update_allergy(patient_id, allergy_id, data)


@router.delete("/{patient_id}/allergies/{allergy_id}", status_code=204)
async def delete_allergy(
    patient_id: uuid.UUID,
    allergy_id: uuid.UUID,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    await PatientService(db, redis).delete_allergy(patient_id, allergy_id)


# ─── Chronic diseases ─────────────────────────────────────────────────────────

@router.post("/{patient_id}/chronic-diseases", response_model=ChronicDiseaseResponse, status_code=201)
async def add_chronic_disease(
    patient_id: uuid.UUID,
    data: ChronicDiseaseCreate,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    return await PatientService(db, redis).add_chronic_disease(patient_id, data)


@router.patch("/{patient_id}/chronic-diseases/{disease_id}", response_model=ChronicDiseaseResponse)
async def update_chronic_disease(
    patient_id: uuid.UUID,
    disease_id: uuid.UUID,
    data: ChronicDiseaseUpdate,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    return await PatientService(db, redis).update_chronic_disease(patient_id, disease_id, data)


@router.delete("/{patient_id}/chronic-diseases/{disease_id}", status_code=204)
async def delete_chronic_disease(
    patient_id: uuid.UUID,
    disease_id: uuid.UUID,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    await PatientService(db, redis).delete_chronic_disease(patient_id, disease_id)


# ─── Documents ────────────────────────────────────────────────────────────────

@router.post("/{patient_id}/documents", response_model=DocumentResponse, status_code=201)
async def upload_document(
    patient_id: uuid.UUID,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
    file: UploadFile = File(...),
):
    return await PatientService(db, redis).upload_document(patient_id, file, uploaded_by=current_user)


@router.get("/{patient_id}/documents", response_model=list[DocumentResponse])
async def get_documents(
    patient_id: uuid.UUID,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    return await PatientService(db, redis).get_documents(patient_id)


@router.delete("/{patient_id}/documents/{document_id}", status_code=204)
async def delete_document(
    patient_id: uuid.UUID,
    document_id: uuid.UUID,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    await PatientService(db, redis).delete_document(patient_id, document_id, deleted_by=current_user)


# ─── History ──────────────────────────────────────────────────────────────────

@router.get("/{patient_id}/history", response_model=list[EncounterSummary])
async def get_history(
    patient_id: uuid.UUID,
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
):
    return await PatientService(db, redis).get_patient_history(patient_id)


# ─── Allergens ────────────────────────────────────────────────────────────────

@router.get("/allergens/search", response_model=list[AllergenResponse])
async def search_allergens(
    current_user: Annotated[User, Depends(require_doctor)],
    db: Annotated[AsyncSession, Depends(get_db)],
    redis: Annotated[aioredis.Redis, Depends(get_redis)],
    q: str = Query(..., min_length=1),
    limit: int = Query(20, ge=1, le=50),
):
    return await PatientService(db, redis).search_allergens(q, limit)

File: frontend/src/api/patients.ts
import client from './client'
import type {
  PatientResponse,
  MedicalCardResponse,
  AllergyResponse,
  ChronicDiseaseResponse,
  DocumentResponse,
  EncounterSummary,
  AllergenResponse,
  Gender,
  BloodType,
  AllergySeverity,
} from './types'

export const patientsApi = {
  list: (params?: { search?: string; skip?: number; limit?: number }) =>
    client.get<PatientResponse[]>('/patients', { params }),

  get: (id: string) => client.get<PatientResponse>(`/patients/${id}`),

  deactivate: (id: string) => client.delete(`/patients/${id}`),

  create: (data: {
    tax_id: string
    unzr?: string
    first_name: string
    last_name: string
    middle_name?: string
    birth_date: string
    gender: Gender
    phone?: string
    email?: string
    address?: Record<string, string>
    primary_doctor_id?: string
    user_email?: string
  }) => client.post<PatientResponse>('/patients', data),

  update: (id: string, data: Partial<{
    first_name: string
    last_name: string
    middle_name: string
    birth_date: string
    gender: Gender
    phone: string
    email: string
    address: Record<string, string>
    unzr: string
  }>) => client.patch<PatientResponse>(`/patients/${id}`, data),

  getMedicalCard: (id: string) =>
    client.get<MedicalCardResponse>(`/patients/${id}/medical-card`),

  updateMedicalCard: (id: string, data: {
    blood_type?: BloodType
    height_cm?: number
    weight_kg?: number
    disability_group?: string
    notes?: string
  }) => client.put<MedicalCardResponse>(`/patients/${id}/medical-card`, data),

  // Allergies
  addAllergy: (id: string, data: {
    substance: string
    severity: AllergySeverity
    reaction?: string
  }) => client.post<AllergyResponse>(`/patients/${id}/allergies`, data),

  updateAllergy: (id: string, allergyId: string, data: {
    substance?: string
    severity?: AllergySeverity
    reaction?: string
  }) => client.patch<AllergyResponse>(`/patients/${id}/allergies/${allergyId}`, data),

  deleteAllergy: (id: string, allergyId: string) =>
    client.delete(`/patients/${id}/allergies/${allergyId}`),

  // Chronic diseases
  addChronicDisease: (id: string, data: {
    icd10_id: string
    diagnosed_at?: string
    notes?: string
  }) => client.post<ChronicDiseaseResponse>(`/patients/${id}/chronic-diseases`, data),

  updateChronicDisease: (id: string, diseaseId: string, data: {
    diagnosed_at?: string
    notes?: string
  }) => client.patch<ChronicDiseaseResponse>(`/patients/${id}/chronic-diseases/${diseaseId}`, data),

  deleteChronicDisease: (id: string, diseaseId: string) =>
    client.delete(`/patients/${id}/chronic-diseases/${diseaseId}`),

  // Documents
  uploadDocument: (id: string, file: File) => {
    const form = new FormData()
    form.append('file', file)
    return client.post<DocumentResponse>(`/patients/${id}/documents`, form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
  },

  getDocuments: (id: string) =>
    client.get<DocumentResponse[]>(`/patients/${id}/documents`),

  deleteDocument: (id: string, documentId: string) =>
    client.delete(`/patients/${id}/documents/${documentId}`),

  getHistory: (id: string) =>
    client.get<EncounterSummary[]>(`/patients/${id}/history`),

  // Allergen reference search
  searchAllergens: (q = '') =>
    client.get<AllergenResponse[]>('/patients/allergens/search', { params: { q } }),
}